Parameterise execution of runner (update)

esteve opened the pull request:
This is just https://github.com/machulav/ec2-github-runner/pull/122 rebased on top of `main` and `dist/index.js` updated after running `npm run package`, all credit goes to @kpturner.

I've also fixed a small issue where if the CI job was set to run as a user, and let it download the GitHub runner code, it wouldn't work (see https://github.com/machulav/ec2-github-runner/pull/122#discussion_r1194972979 for context, fix is in https://github.com/machulav/ec2-github-runner/commit/35580d0afba9bc09a2e504ba34491d7079408e53)

esteve commented:
@kpturner  I hope it's ok with you I submitted this PR, I don't mean to take credit for your work, I've submitted it in the hopes that it'd be easier for @machulav to merge your work more quickly (`dist/index.js` up to date, fix for runner running as a user).

kpturner commented:
> @kpturner I hope it's ok with you I submitted this PR, I don't mean to take credit for your work, I've submitted it in the hopes that it'd be easier for @machulav to merge your work more quickly (`dist/index.js` up to date, fix for runner running as a user).

It is fine by me - as long as it helps somebody I am not bothered about who gets the credit :)

esteve commented:
@machulav I've rebased this PR on top of `main` to fix the conflicts. Is there anything that needs to be addressed before being merged? Thanks.

alonbl commented on src/aws.js:
```diff
@@ -27,9 +29,18 @@ function buildUserDataScript(githubRegistrationToken, label) {
+    userData.push(`./svc.sh install ${config.input.runAsUser || ''}`);
+    userData.push('./svc.sh start');
+  } else {
+    userData.push(`${config.input.runAsUser ? `su ${config.input.runAsUser} -c` : ''} ./run.sh`);
```
if `run.sh` supports only running as root then why do we run it under regular user? and if it supported the description of `run-runner-as-user` parameter should be updated.

and if this is possible why do we need to run it as a service?

also, won't it better to create the user within this script? also add it to sudoers to be closer to github environment?

hvgazula commented:
Hello! Any chance this PR will be merged to enable running the ec2 self-hosted runner as non-root?

esteve commented:
@machulav @hvgazula we've been using this branch for the past eight months at the Autoware Foundation without an issue, but I'd prefer it to get merged instead of using our fork. Is there any feedback I should address? Thanks.

hvgazula commented:
> @machulav @hvgazula we've been using this branch for the past eight months at the Autoware Foundation without an issue, but I'd prefer it to get merged instead of using our fork. Is there any feedback I should address? Thanks.

Thank you for replying @esteve. Is it on the main branch of your fork? I agree it'd be nice to have this feature merged in here.  I'll ask this again for attention :). @machulav any chance you can merge this feature?

esteve commented:
@hvgazula it's here https://github.com/esteve/ec2-github-runner/tree/Optionally-execute-runner-as-a-service

My fork is just an updated version of @kpturner 's work from https://github.com/machulav/ec2-github-runner/pull/122 plus a couple of minor fixes needed to run as a non-root user.

hvgazula commented:
Thanks, @esteve. It worked.

thimios commented:
@machulav we also need to run specs as a non-root user, it would be great to see this merged

Preen commented:
@esteve and @kpturner can you resolve the conflicts, and I will test it out and merge it into main asap.

esteve commented:
@Preen sorry, I somehow missed your comment from months ago, I'll update the PR. Thanks.

esteve commented:
@Preen I've fixed the conflicts and ran `npm run package` to update `dist/index.js`, let me know if there's more feedback to address. However, I know next to nothing about Javascript, I merely updated @kpturner 's work and fixed a couple of bugs, so feel free to push any changes you think this PR needs. Thanks!

lifeisafractal commented:
@esteve Thanks for keeping up on this one. I'd love to see this merged into this repo so I can go back to using the main upstream. FYI, I was pointing at your branch and the recent rebase seems to have borken things (I opened a PR to your fork just to highlight it).

esteve commented:
@lifeisafractal I've just merged your changes, thanks! 🙂

esteve commented:
@lifeisafractal I ran `npm run package`, but looks like I must be missing something. This is what I did before your fix:

```sh
docker run -e NODE_OPTIONS=--openssl-legacy-provider -it --rm -v `pwd`:`pwd` -w `pwd` node sh -c "npm run package"
```

I'm working on adding an option to set the root volume size in another branch, but after running `npm run package`, it seems to be broken. Is there anything I should be running instead? Thanks.

lifeisafractal commented:
@esteve I'm not far (if at all) ahead of you with Node. I did a fresh clone with a `npm install` followed by `npm run package` and that did the trick. I don't know too much more than that Glad you got things working!

Preen commented:
Tested successfully, merged!
